File: __tests__/unit/test_smart_preprocessor.ts
import { describe, it, expect, beforeEach } from 'vitest';
import { SmartPreprocessor } from 'ztb/evaluation/preprocess';
import { CommonPreprocessor } from 'ztb/features/base';
import * as pd from 'pandas-js';

describe('SmartPreprocessor', () => {
  let preprocessor: SmartPreprocessor;
  let sampleData: any;

  beforeEach(() => {
    preprocessor = new SmartPreprocessor();

    // Create sample OHLC data
    const dates = pd.date_range('2020-01-01', 100, 'D');
    sampleData = new pd.DataFrame({
      'date': dates.toArray(),
      'open': Array.from({length: 100}, () => 100 + Math.random() * 10),
      'high': Array.from({length: 100}, () => 105 + Math.random() * 10),
      'low': Array.from({length: 100}, () => 95 + Math.random() * 10),
      'close': Array.from({length: 100}, () => 100 + Math.random() * 10),
      'volume': Array.from({length: 100}, () => Math.floor(Math.random() * 10000) + 1000)
    });
    const preprocessedData = CommonPreprocessor.preprocess(sampleData);
    sampleData = preprocessedData;
    sampleData = CommonPreprocessor.preprocess(sampleData);
  });

  describe('preprocess', () => {
    it('should calculate only required ema columns', () => {
      const required = ['ema:12', 'ema:26'];
      const result = preprocessor.preprocess(sampleData, required);

      expect(result.columns).toContain('ema_12');
      expect(result.columns).toContain('ema_26');
      expect(result.columns).not.toContain('ema_5');
      expect(result.columns).not.toContain('ema_50');
    });

    it('should calculate only required rolling columns', () => {
      const required = ['rolling:10:mean', 'rolling:20:std'];
      const result = preprocessor.preprocess(sampleData, required);

      expect(result.columns).toContain('rolling_mean_10');
      expect(result.columns).toContain('rolling_std_20');
      expect(result.columns).not.toContain('rolling_mean_5');
      expect(result.columns).not.toContain('rolling_std_30');
    });

    it('should handle mixed requirements', () => {
      const required = ['ema:12', 'rolling:20:mean', 'close'];
      const result = preprocessor.preprocess(sampleData, required);

      expect(result.columns).toContain('ema_12');
      expect(result.columns).toContain('rolling_mean_20');
      expect(result.columns).toContain('close');
      expect(result.columns).not.toContain('ema_26');
      expect(result.columns).not.toContain('rolling_std_20');
    });

    it('should handle empty requirements', () => {
      const required: string[] = [];
      const result = preprocessor.preprocess(sampleData, required);

      // Should still include basic columns
      expect(result.columns).toContain('close');
      expect(result.columns).toContain('open');
      expect(result.columns).toContain('high');
      expect(result.columns).toContain('low');
      expect(result.columns).toContain('volume');
    });

    it('should cache calculations', () => {
      const required = ['ema:12', 'ema:26'];

      // First call
      const result1 = preprocessor.preprocess(sampleData, required);

      // Second call with same requirements should use cache
      const result2 = preprocessor.preprocess(sampleData, required);
      expect(result1.to_json()).toBe(result2.to_json());
      expect(result1.equals(result2)).toBe(true);
    });
  });
});

File: __tests__/unit/test_smart_preprocessor.py
#!/usr/bin/env python3
"""
Unit tests for SmartPreprocessor
"""

import unittest

import numpy as np
import pandas as pd

from ztb.evaluation.preprocess import SmartPreprocessor


class TestSmartPreprocessor(unittest.TestCase):
    """Test cases for SmartPreprocessor"""

    def setUp(self):
        """Set up test fixtures"""
        # Create sample OHLC data (without CommonPreprocessor.preprocess)
        dates = pd.date_range("2020-01-01", periods=100, freq="D")
        np.random.seed(42)
        self.sample_data = pd.DataFrame(
            {
                "open": 100 + np.random.randn(100).cumsum(),
                "high": 105 + np.random.randn(100).cumsum(),
                "low": 95 + np.random.randn(100).cumsum(),
                "close": 100 + np.random.randn(100).cumsum(),
                "volume": np.random.randint(1000, 10000, 100),
            },
            index=dates,
        )

    def test_preprocess_ema_only(self):
        """Test preprocessing with only EMA requirements"""
        required = ["ema:12", "ema:26"]
        preprocessor = SmartPreprocessor(set(required))
        result = preprocessor.preprocess(self.sample_data)

        # Should contain required EMA columns
        self.assertIn("ema_12", result.columns)
        self.assertIn("ema_26", result.columns)

        # Should not contain other EMA columns
        self.assertNotIn("ema_5", result.columns)
        self.assertNotIn("ema_50", result.columns)

        # Should contain basic columns
        self.assertIn("close", result.columns)
        self.assertIn("open", result.columns)

    def test_preprocess_rolling_only(self):
        """Test preprocessing with only rolling requirements"""
        required = ["rolling_mean:10", "rolling_std:20"]
        preprocessor = SmartPreprocessor(set(required))
        result = preprocessor.preprocess(self.sample_data)

        # Should contain required rolling columns
        self.assertIn("rolling_mean_10", result.columns)
        self.assertIn("rolling_std_20", result.columns)

        # Should not contain other rolling columns
        self.assertNotIn("rolling_mean_5", result.columns)
        self.assertNotIn("rolling_std_30", result.columns)

    def test_preprocess_mixed_requirements(self):
        """Test preprocessing with mixed requirements"""
        required = ["ema:12", "rolling_mean:20", "close"]
        preprocessor = SmartPreprocessor(set(required))
        result = preprocessor.preprocess(self.sample_data)

        # Should contain all required columns
        self.assertIn("ema_12", result.columns)
        self.assertIn("rolling_mean_20", result.columns)
        self.assertIn("close", result.columns)

        # Should not contain non-required columns
        self.assertNotIn("ema_26", result.columns)
        self.assertNotIn("rolling_std_20", result.columns)

    def test_preprocess_empty_requirements(self):
        """Test preprocessing with empty requirements"""
        required = []
        preprocessor = SmartPreprocessor(set(required))
        result = preprocessor.preprocess(self.sample_data)

        # Should still include basic columns
        self.assertIn("close", result.columns)
        self.assertIn("open", result.columns)
        self.assertIn("high", result.columns)
        self.assertIn("low", result.columns)
        self.assertIn("volume", result.columns)

    def test_preprocess_caching(self):
        """Test that preprocessing caches calculations"""
        required = ["ema:12", "ema:26"]
        preprocessor = SmartPreprocessor(set(required))

        # First call
        result1 = preprocessor.preprocess(self.sample_data)

        # Second call with same requirements should use cache
        result2 = preprocessor.preprocess(self.sample_data)

        # Results should be identical
        pd.testing.assert_frame_equal(result1, result2)

    def test_preprocess_invalid_requirement(self):
        """Test preprocessing with invalid requirement format"""
        required = ["invalid_format", "ema:12"]
        preprocessor = SmartPreprocessor(set(required))
        result = preprocessor.preprocess(self.sample_data)

        # Should still process valid requirements
        self.assertIn("ema_12", result.columns)

        # Should contain basic columns
        self.assertIn("close", result.columns)


if __name__ == "__main__":
    unittest.main()
